packages/common/src/subchain/SubchainClient.ts

```diff
@@ -0,0 +1,142 @@
+        tokenAccount: string,
+        atomicAccount: string,
+        atomicmarketAccount: string,
+        wasmResponse: PromiseLike<Response>,
+        stateResponse: PromiseLike<Response>,
+        blocksUrl: string,
+        slowmo = false
+    ) {
+        this.blocksUrl = blocksUrl;
+        this.slowmo = slowmo;
+        if (this.shuttingDown) return this.shutdown();
+        const [, state] = await Promise.all([
+            this.subchain.instantiateStreaming(wasmResponse),
+            stateResponse.then((resp) => {
+                if (resp.ok) return resp.arrayBuffer();
+                return null;
+            }),
+        ]);
+        if (this.shuttingDown) return this.shutdown();
+        if (state) {
+            this.subchain.setMemory(state);
+        } else {
+            this.subchain.initializeMemory(
+                edenAccount,
+                tokenAccount,
+                atomicAccount,
+                atomicmarketAccount
+            );
+        }
+        this.connect();
+    }
+
+    connect() {
+        if (this.shuttingDown) return this.shutdown();
+        this.ws = new WebSocket(this.blocksUrl);
+        this.ws.onclose = () => {
+            console.error("Closed connection to " + this.blocksUrl);
+            this.ws = null;
+            if (!this.shuttingDown) setTimeout(() => this.connect(), 500);
+        };
+        this.ws.onerror = () => {
+            console.error("Error connecting to " + this.blocksUrl);
+            this.ws = null;
+        };
+        this.ws.onopen = () => {
+            console.log("Connected to " + this.blocksUrl);
+            if (this.slowmo) {
+                setTimeout(() => this.sendStatus(), 1000);
+            } else {
+                this.sendStatus();
+            }
+        };
+        this.ws.onmessage = (ev: MessageEvent<any>) => this.onmessage(ev);
+    }
+
+    async sendStatus() {
+        if (this.shuttingDown) return this.shutdown();
+        if (!this.ws) return;
+        const irreversible: number = this.subchain.getIrreversible();
+        const blocks: BlockInfo[] = this.subchain
+            .query(
+                `{blockLog{blocks(ge:${irreversible}){edges{node{num id}}}}}`
+            )
+            .data.blockLog.blocks.edges.map((e: any) => e.node);
+        const stat: ClientStatus = {
+            maxBlocksToSend: this.slowmo ? 1 : 1000,
+            irreversible,
+            blocks,
+        };
+        if (this.slowmo)
+            setTimeout(() => this.ws?.send(JSON.stringify(stat)), 50);
+        else this.ws.send(JSON.stringify(stat));
+    }
+
+    messageQueue: (Promise<ArrayBuffer> | string)[] = [];
+    processingQueue = false;
+
+    async processQueue() {
+        if (this.processingQueue) return;
+        this.processingQueue = true;
+        try {
+            while (this.messageQueue.length) {
+                if (this.shuttingDown) return this.shutdown();
+                const data = this.messageQueue.shift();
```

Comment:
should we be using a proper queue structure as opposed to js array to have O(1) dequeue operations? Or this would be preoptimization at this time? (it can also be in a later pr)

Reply:
Here's a little secret that most people don't know: vector (array in js) operations are often faster because of the way the CPU's memory cache operates (it favors linear scans). dequeues sometimes have an advantage, but usually only for large n (and even then, not always for large n), which shouldn't happen here.

Reply:
The performance difference is generally not enough to overcome a big-O difference.  As long as the queue stays empty, you'll be fine, but if it falls too far behind, you might never catch up.  A deque is still reasonably cache-friendly, but is a little slower due to adding one level of indirection.  The most common optimized implementation of a queue is a circular buffer which has the same performance characteristics as vector.

Reply:
got it! thanks for the both thorough explanations! I was not sure how big this queue would be